Add support for STATES of vacuums

cnrd opened the pull request:
## Description:
Frontend changes: https://github.com/home-assistant/home-assistant-polymer/pull/1497

Added support for states as agreed upon in: https://github.com/home-assistant/architecture/issues/29

The xiaomi_miio platform has been updated to support the new states as part of this PR.

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

cnrd committed:
Vacuum: Added support for STATES

rytilahti commented on homeassistant/components/vacuum/xiaomi_miio.py:
```diff
@@ -82,6 +84,20 @@
+    15: STATE_RETURNING,
+    9: STATE_ERROR,
+    12: STATE_ERROR,
+}
```
Would it make sense to order this by the key, so that one would see what items are missing? e.g. 4 is "remote control active" or 7 is "manual mode", which are missing here, see https://github.com/rytilahti/python-miio/blob/master/miio/vacuumcontainers.py#L53 .

Furthermore it may be useful to (at least debug) log the missing state code, maybe even print out the state and mention that it is not handled?

cnrd committed:
Added debug logging and corrected state order

cnrd committed:
typo

cnrd committed:
Fix travis error, STATE = STATE for readability

cnrd committed:
status -> state

awarecan commented on homeassistant/components/vacuum/__init__.py:
```diff
@@ -217,8 +195,8 @@ def supported_features(self):
     @property
-    def status(self):
-        """Return the status of the vacuum cleaner."""
+    def state(self):
```
You rewrite `state` property of `ToggleEntity`?

cnrd replied:
You are right, it should no longer be a `ToggleEntity` as we are removing the toggle.

awarecan reviewed (changes requested):
This PR breaks all `vacuum`

Either change it to a non-break way to implement new feature, and only include base class and demo platform in PR.

Or fix all platforms in one PR.

cnrd committed:
Changed to Entity instead of ToogleEntity

cnrd committed:
Updated some vacuums

awarecan commented:
Your title is 'Added support for STATES', but what you done is change STATUS -> STATES, that will break too many platforms. I would suggest you just **add** STATES, keep old STATUS feature aside, and allow each platform retire STATUS feature one by one.

awarecan commented:
Sounds good. That is what supported features meant to.

cnrd committed:
Revert changes

cnrd committed:
Revert Changes

cnrd committed:
added SUPPORT_STATE

cnrd committed:
Woof?

cnrd committed:
Implement on/off if STATE not supported

cnrd commented:
This should make for a better PR.

I changed it so that it hopefully will be backwards compatible and I moved the Xiaomi Vacuum platform to it's own PR.

awarecan commented:
Thanks for your quickly action. Please change demo vacuum for your new design and include it into this PR. So that we can know how the new design works.

Also add test for the demo vacuum.

cnrd commented:
Not sure if you saw the edit:
Should I keep it as ToggleEntity or should I make it an Entity? As it seems like I would have to reimplement a lot of the stuff from ToggleEntity if I change it to Entity.

c727 commented:
I think the states for this should be discussed in https://github.com/home-assistant/architecture first.
`SUPPORT_STATE` is a bad name for a feature

cnrd commented:
Already done: https://github.com/home-assistant/architecture/issues/29

cnrd committed:
Moved new state vaccum to Class StateVacuumDevice

cnrd committed:
Error: I should go to bed

cnrd committed:
Moved around methods for easier reading

cnrd committed:
Added StateVacuumDevice demo vacuum

cnrd committed:
Added tests for StateVacuumDevice demo vacuum

cnrd committed:
Fix styling errors

cnrd commented:
I changed it such that states are in a new class, which avoids touching old platforms at all.
I added a demo vacuum and added tests for the demo vacuum.
I'm pretty happy with the current code.
@awarecan ping if you want to look though it?

awarecan commented on homeassistant/components/vacuum/demo.py:
```diff
@@ -204,3 +212,111 @@ def send_command(self, command, params=None, **kwargs):
+        if self.supported_features & SUPPORT_RETURN_HOME == 0:
+            return
+
+        self._state = STATE_RETURNING
```
How this device can back to dock?

You can mock this behavior by using `hass.loop.call_later(30, set_state_to_dock)`

awarecan reviewed (changes requested):
I think you can refactor it a little more.

```
BaseVacuum                      ToggleEntitiy
 |      \                           /
 |       \--------VacuumDevice ----/
 |
 |-----StateVacuumDevice
```

cnrd commented:
I'm not sure I understand your drawing.

Should I create a BaseVacuum that uses BaseVacuum.StateVacuumDevice and BaseVacuum.VacuumDevice depending on supported_features?

cnrd committed:
Refactored to BaseVaccum

cnrd committed:
Vacuum will now go back to dock

cnrd commented:
@awarecan I'm pretty sure I did everything you asked for except the `**kwargs` thing as that would require me to change all vacuum platforms, which is what we are trying to avoid. This can be done in another PR.

cnrd committed:
Class BaseVacuum is for internal use only

cnrd commented:
@awarecan any more changes needed or are we good to go?

awarecan commented:
I am okay with this change. @balloob would you mind look at it?

MartinHjelmare commented on homeassistant/components/vacuum/__init__.py:
```diff
@@ -245,45 +242,28 @@ def fan_speed_list(self):
         This method must be run in the event loop and returns a coroutine.
         """
-        return self.hass.async_add_job(partial(self.turn_on, **kwargs))
+        return self.hass.async_add_executor_job(
```
This method should not return anything. Instead `await` the task returned by `self.hass.async_add_executor_job`.

MartinHjelmare replied:
Also update the docstring to reflect the change.

MartinHjelmare reviewed (commented):
Please update the async actuator methods. See below.

cnrd committed:
return -> await

cnrd committed:
return -> await

cnrd commented:
@MartinHjelmare Is that what you wanted? :-)

MartinHjelmare commented:
Yes!

arbreng commented on homeassistant/components/vacuum/demo.py:
```diff
@@ -204,3 +212,118 @@ def send_command(self, command, params=None, **kwargs):
+
+    def start_pause(self, **kwargs):
+        """Start, pause or resume the cleaning task."""
+        if self.supported_features & SUPPORT_PAUSE == 0:
```
Doesn't this mean the vacuum cannot be started if it doesn't support pause?  The demo vacuum doesn't support pause so it should definitely be broken.

Suggest doing this below inside of if self._state == STATE_CLEANING: ...

However, this means that calling this method on a CLEANING vacuum which doesn't support pause will just do nothing.  Is that OK?

cnrd replied:
True, I wanted to do a refactor of start_pause for all vacuums (splitting them up) in a separate PR, seeing as this currently isn't a problem (all current platforms support pause), this could be added in another platform in just a couple of lines of code.

arbreng replied:
Oh, SUPPORT_STATE_SERVICES contains pause.  I'm way too tired.

cnrd commented:
I added a PR for the start_pause split-up, because as @arbreng pointed out, if there is ever a Vacuum that does not support pausing, there is no way to turn it on in the `StateVacuumDevice` I'm not sure if I did the PR correctly, as I build it on this branch and it therefor includes all of the code from this PR on top of the changes to start_pause.
